### ISSUE TITLE
react-leaflet Components cannot be wrapped in React components

### ISSUE THREAD

**Issue by uniphil** [PaulLeCam/react-leaflet#73]
In apps I work on, we often want to do things like:

- Have different views over the same map (maybe controlled by react-router)
- Share customized `Marker` wrapper code between different views.

For both of these use-cases we should be able to make new React components that can be placed on the map, containing only react-leaflet children.

Example code:

```jsx
import React, { Component } from 'react';
import { Map, Marker } from 'react-leaflet';

class WrappedMarker extends Component {
  render() {
    return <Marker position={[51.505, -0.09]} />;
  }
}

export default class MapView extends Component {
  render() {
    return (
      <Map center={[51.505, -0.09]} zoom={13}>
        <WrappedMarker />
      </Map>
    );
  }
}
```

The above does not show the marker, and logs a warning:
```
Uncaught TypeError: Cannot read property 'addLayer' of undefined
```

**Comment by uniphil** [comment]
Since [`React.Children.map`](https://facebook.github.io/react/docs/top-level-api.html#react.children.map) only does a shallow mapping, this cannot be handled automatically.

The wrapping component can pass `this.props.map` down to its child manually and it works.

**Comment by Smallmelo** [comment]
@uniphil i'm running into this same problem, and if you closed this ticket because you found a solution, it is unclear to me what it was. can you elaborate?

**Comment by jgimbel** [comment]
[Here](https://github.com/PaulLeCam/react-leaflet/blob/master/example/components/custom-component.js) is an example of how to create custom components, and the [documentation](https://github.com/PaulLeCam/react-leaflet#creating-custom-components). 

TLDR; Notice [line 5](https://github.com/PaulLeCam/react-leaflet/blob/master/example/components/custom-component.js#L5). When creating a custom component, the `map` and `layerContainer` props must be passed down.

The initial example can be fixed by adding these props to the `Marker`.

```js
import React, { Component } from 'react';
import { Map, Marker } from 'react-leaflet';

class WrappedMarker extends Component {
  render() {
    const { map, layerContainer } = this.props; //Given by the `Map` component
    return (
      <Marker 
        map={map} /* pass down to Marker */
        layerContainer={layerContainer} /* pass down to Marker */
        position={[51.505, -0.09]} 
      />
    );
  }
}

export default class MapView extends Component {
  render() {
    return (
      <Map center={[51.505, -0.09]} zoom={13}>
        <WrappedMarker />
      </Map>
    );
  }
}
```

**Comment by Smallmelo** [comment]
thank you, @jgimbel.  that cleared things up. makes much more sense.

**Comment by mdfischer** [comment]
I have run into the same problem with the MapControl react-leaflet component, and the workaround by @jgimbel above does not solve it in this case.  As far as I understand from the release notes of react-leaflet v0.12.0 (https://github.com/PaulLeCam/react-leaflet/blob/2853dff5192c40f7715e6330e25c45d0bc398008/UPGRADING.md) this workaround shouldn't even be necessary anymore anyway, because the of the use of React context.  (I've confirmed that's the case for the TileLayer and WMSTileLayer components, which work fine when wrapped in custom React components).

**Comment by taozhi8833998** [comment]
in react-leaflet 1.1.1version, the map and layerContainer is not in props of child component

**Comment by cbrizuelaromero** [comment]
> Attempted import error: 'Map' is not exported from 'react-leaflet' (imported as 'LeafletMap').

I changed 'Map' to 'MapContainer' in the import:

_import { MapContainer as LeaftletMap, TileLayer } from 'react-leaflet';_

**Comment by leilabaneshi** [comment]
> > Attempted import error: 'Map' is not exported from 'react-leaflet' (imported as 'LeafletMap').
> 
> I changed 'Map' to 'MapContainer' in the import:
> 
> _import { MapContainer as LeaftletMap, TileLayer } from 'react-leaflet';_

thanks it works for me

**Comment by helinagirmay** [comment]
Attempted import error: 'MapContainer' is not exported from 'react-leaflet'....getting this error

**Comment by cbrizuelaromero** [comment]
> Attempted import error: 'MapContainer' is not exported from 'react-leaflet'....getting this error

which version do you use?

**Comment by ashutoshm1771** [comment]
Use 'MapContainer' instead of 'Map'

**Comment by MichelMuemboIlunga** [comment]
After changed from 'Map' to 'MapContainer' in the import I got this error

C:/Users/Private/node_modules/@react-leaflet/core/esm/index.js
SyntaxError: C:\Users\Private\node_modules\@react-leaflet\core\esm\index.js: Unexpected character '' (1:0)

> 1 |
    | ^
    at parser.next (<anonymous>)

**Comment by TimMcCauley** [comment]
I had a similar issue and came up with this solution https://codesandbox.io/s/react-leaflet-overlay-import-wu6ct?file=/src/index.js:333-704

**Comment by nirmalyathakurta** [comment]
> After changed from 'Map' to 'MapContainer' in the import I got this error
> 
> C:/Users/Private/node_modules/@react-leaflet/core/esm/index.js
> SyntaxError: C:\Users\Private\node_modules@react-leaflet\core\esm\index.js: Unexpected character '' (1:0)
> 
> > 1 |
> > | ^
> > at parser.next ()

Having the same issue after updating map to mapcontainer
